service/history/timerQueueTaskExecutorBase.go

```diff
@@ -91,17 +95,34 @@ func (t *timerQueueTaskExecutorBase) executeDeleteHistoryEventTask(
 	defer func() { release(retError) }()
 
 	mutableState, err := loadMutableStateForTimerTask(weContext, task, t.metricsClient, t.logger)
-	if err != nil {
+	switch err.(type) {
+	case nil:
+		// continue the following step
+	case *serviceerror.NotFound:
+		// the mutable state is deleted and delete history branch operation failed.
+		// use task branch token to delete the leftover history branch
+		return t.deleteHistoryBranch(task.BranchToken)
+	default:
 		return err
 	}
 
 	lastWriteVersion, err := mutableState.GetLastWriteVersion()
 	if err != nil {
 		return err
 	}
-	ok, err := verifyTaskVersion(t.shard, t.logger, namespace.ID(task.NamespaceID), lastWriteVersion, task.Version, task)
-	if err != nil || !ok {
-		return err
+	if ok := verifyTaskVersion(t.shard, t.logger, mutableState.GetNamespaceEntry(), lastWriteVersion, task.Version, task); !ok {
+		currentBranchToken, err := mutableState.GetCurrentBranchToken()
+		if err != nil {
+			return err
+		}
+		// the mutable state has a newer version and the branch token is updated
+		// use task branch token to delete the original branch
+		if !bytes.Equal(task.BranchToken, currentBranchToken) {
+			return t.deleteHistoryBranch(task.BranchToken)
+		}
+		// it is not expected to hit here. Usually, an updated version means a different branch token
+		// this is to prevent us from deleting the correct branch token
+		return nil
```

Comment:
if this case is not expected and can be a potential bug, probably emit some log or metric so we know it happens?